src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -165,22 +165,27 @@ async ValueTask<(Solution? solution, T result)> TryFastGetSolutionAndRunAsync()
         {
             // See if anyone else is computing this solution for this checksum.  If so, just piggy-back on that.  No
             // need for us to force the same computation to happen ourselves.
+            var currentSolution = this.CurrentSolution;
             var lazySolution = await GetLazySolutionAndIncrementRefCountAsync().ConfigureAwait(false);
 
             try
             {
                 // Actually get the solution, computing it ourselves, or getting the result that another caller was computing.
-                var solution = await lazySolution.GetValueAsync(cancellationToken).ConfigureAwait(false);
+                var newSolution = await lazySolution.GetValueAsync(cancellationToken).ConfigureAwait(false);
+
+                // We may have just done a lot of work to determine the up to date primary branch solution.  See if we
+                // can move the workspace forward to that solution snapshot.
+                (newSolution, _) = await this.TryUpdateWorkspaceAsync(
+                    workspaceVersion, fromPrimaryBranch, newSolution, cancellationToken).ConfigureAwait(false);
```

Comment:
see comment below.  Now, we ensure that all concurrent calls can get the large computed solution, but then calls for hte 'fromPrimaryBranch' will always try to at least update the workspace if possible here.  Only the first will succeed as we only move the workspace version forward, so any other concurrent calls will see that it happened, and will bail out immediately.